### PR TITLE
Create `matrix.h` and `matrix_stack.h`

### PR DESCRIPTION
**Pull request by Dragorn421**
Create `matrix_stack.h` with all the stack-based matrix manipulation functions, and `matrix.h` with the rest of the matrix functions (besides the `gu` ones which should be in libultra's gu.h (and currently are in functions.h))

### CODE REVIEW

File: include/matrix.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+void SkinMatrix_Vec3fMtxFMultXYZW(MtxF* mf, Vec3f* src, Vec3f* xyzDest, f32* wDest);
+void SkinMatrix_Vec3fMtxFMultXYZ(MtxF* mf, Vec3f* src, Vec3f* dest);
+void SkinMatrix_MtxFMtxFMult(MtxF* mfA, MtxF* mfB, MtxF* dest);
+void SkinMatrix_GetClear(MtxF** mfp);
+void SkinMatrix_MtxFCopy(MtxF* src, MtxF* dest);
+s32 SkinMatrix_Invert(MtxF* src, MtxF* dest);
+void SkinMatrix_SetScale(MtxF* mf, f32 x, f32 y, f32 z);
+void SkinMatrix_SetRotateZYX(MtxF* mf, s16 x, s16 y, s16 z);
+void SkinMatrix_SetTranslate(MtxF* mf, f32 x, f32 y, f32 z);
+void SkinMatrix_SetTranslateRotateYXZScale(MtxF* dest, f32 scaleX, f32 scaleY, f32 scaleZ, s16 rotX, s16 rotY, s16 rotZ,
```
Comment (fig02) [inline include/matrix.h:80]: I'd think it makes more sense for sys_matrix and skin_matrix to be separate headers. They are different systems that aren't really compatible with each other, being that one manipulates the matrix stack and the other does not.

Take for example eff_ss_kirakira, which opts to use solely skinmatrix operations instead of the matrix stack. It should probably just be able to include the specific matrix system it wants to use on its own

Comment (Dragorn421) [reply include/matrix.h:80]: What do we gain from two separate headers?

I think we should split headers by feature rather than make basically one .h per .c, otherwise we'd end up with separate headers for the same thing (like matrix operations here) or a single header with way too much going on (z_actor.c)

Do note that sys_matrix.c is not only operations on the matrix stack, it also has operations that don't involve the stack:
```c
Mtx* Matrix_MtxFToMtx(MtxF* src, Mtx* dest);
void Matrix_MtxFCopy(MtxF* dest, MtxF* src);
void Matrix_MtxToMtxF(Mtx* src, MtxF* dest);
void Matrix_MultVec3fExt(Vec3f* src, Vec3f* dest, MtxF* mf);
void Matrix_Transpose(MtxF* mf);
void Matrix_MtxFToYXZRotS(MtxF* mf, Vec3s* rotDest, s32 flag);
void Matrix_MtxFToZYXRotS(MtxF* mf, Vec3s* rotDest, s32 flag);
MtxF* Matrix_CheckFloats(MtxF* mf, const char* file, int line);
void Matrix_SetTranslateScaleMtx2(Mtx* mtx, f32 scaleX, f32 scaleY, f32 scaleZ, f32 translateX, f32 translateY,
                                  f32 translateZ);
```

I would be ok splitting the header between "matrix_stack.h" and "matrix_stackless.h" (?), but those `Matrix_` functions should then still go with `SkinMatrix_` functions into "matrix_stackless.h" imo

Comment (fig02) [reply include/matrix.h:80]: Yeah I am not suggesting one header per c file, but one header per system. These are completely different systems that dont work with each other. Whereas yes, z_actor.c contains a lot of different systems in it, so not all of it will be in the same header.

We gain the granularity to allow the programmer to pick which system they want to include into their file. I dont think you should have to include a bunch of unnecessary things because you want to work with the concept of "matrices" in general.

--

Sure, we could look into some scenario where the matrix stack operations in sys_matrix belong to a new system, and you would have functions like `MatrixStack_Init`, `MatrixStack_Push`, `MatrixStack_Pop`... etc. This would just be a huge improvement anyway, and let the user know point blank that they are messing with the stack. 

In this scenario we could also look into changing the `SkinMatrix_` system to just `Matrix_`, along with the other generic matrix operations you mention in sys_matrix.c. This aligns with how things are used anyway-- skin matrix functions clearly started out as helpers for the skin skeleton system, but other devs found it useful outside of that system. It makes sense for us to name it generically to suit that.

**Review by fig02 — APPROVED** [review]
like I mentioned in discord, not sold on the exact split of things here. but we can get back to this this at a later time.

I would almost prefer reverting back to the generic matrix.h as our temporary solution. but whatever others want

File: include/matrix.h (hunk 1 above)
Comment (cadmic) [reply include/matrix.h:45]: I also think it should be `sys_matrix.h` and `skin_matrix.h`. To it seems like "matrix.h" and "matrix_stack.h" is the same thing, except with much worse names, because the names don't match the subsystems they correspond to

Comment (Dragorn421) [reply include/matrix.h:45]: Which subsystems? My point is we shouldn't split headers like the .c files are
At least if the goal is usability of the headers. Which is my goal since I don't see another one

Comment (fig02) [reply include/matrix.h:45]: Nobody is arguing for every c file has its own header, but in this case the two systems are split up into their own c files.

@cadmic they are not quite the same. noted above, there are some sys_matrix functions that do not deal with the matrix stack, and got included in matrix.h, not matrix_stack.h
Im not arguing for keeping this split this way, but just clarifying that they are not the same approach

Comment (cadmic) [reply include/matrix.h:45]: I see, in that case I think my preferences are
sys_matrix.h+skin_matrix.h > matrix.h > matrix.h+matrix_stack.h

Comment (fig02) [reply include/matrix.h:45]: I would agree. I think the matrix_stack approach could work with some eventually with some work on names. but it would be really confusing right now

Comment (Dragorn421) [reply include/matrix.h:45]: When you say "system" do you mean a group of functions that are centered on a specific topic, or the prefix in function names like "Matrix" in `Matrix_Put` ?

If we rename matrix_stack functions to MatrixStack_ and other functions to Matrix_ would things be fine as is then?

Comment (fig02) [reply include/matrix.h:45]: Its not just about a specific concept like "matrices", but a specific implementation of something. A group of functions that make up a system to make something work. 
You can see with sys_matrix and skin_matrix they are completely separate implementations with the same goal. There are parallels across both systems, functions that do the same thing in a different implementation because they are separate. Some examples:
```
Matrix_MtxFCopy - SkinMatrix_MtxFCopy
Matrix_MtxToMtxF - SkinMatrix_MtxFToMtx
Matrix_Scale - SkinMatrix_SetScale
Matrix_RotateZYX - SkinMatrix_SetRotateZYX
Matrix_TranslateRotateZYX - SkinMatrix_SetTranslateRotateZYX
```

RE your point about `SystemName_` in functions-- the name is an indicator of the system, yeah. Youll find all over that the system name of a group of functions does not necessarily have to match that of the c file they are contained in. See `BodyBreak_` contained within z_actor.c as an example. It should indicate a group of functions that implement a certain system. Im not a huge object oriented programming guy, but maybe its appropriate to see this as a class.

Just throwing `MatrixStack_` on the stack-related functions doesn't just fix the problem, if people don't agree that this is where the line is drawn for the different systems. I think the stack separation could maybe work, but ultimately think sys_matrix vs skin_matrix, as the developers have already separated them, makes sense as separate systems.

Comment (cadmic) [reply include/matrix.h:45]: Not sure if you were asking me or fig, but personally I mean something close to "group of functions that are centered on a specific topic" but it's not quite that, since here there are two different sets of functions on the topic of "matrices" but it's still clear they were developed separately and don't behave as a unified system. In a well-designed program systems should correspond to topics but things aren't very well-designed here.

I agree `MatrixStack_` names would be better (though probably in a different PR) but even in that case I think I would still prefer sys_matrix.h+skin_matrix.h > matrix.h > matrix.h+matrix_stack.h (but I'd have to think about it)

Comment (cadmic) [reply include/matrix.h:45]: (also I guess it should be z64skin_matrix.h?)

Comment (fig02) [reply include/matrix.h:45]: Agree, if this were well designed we wouldnt be having this argument and there would be one matrix system that does it all and every developer on the game used that one system. But that is not what happened. Someone made their own matrix system implementation for the skin horse skeletons, and then more devs later decided to use those functions *outside* of the overarching skin system.

Comment (Dragorn421) [reply include/matrix.h:45]: So the reasoning for wanting sys_matrix.h and skin_matrix.h is that that corresponds to the .c files right?

Comment (Dragorn421) [reply include/matrix.h:45]: To be clear with the goal of headers being usable, having two distinct "matrix things" headers would just be confusing, with names as vague as "skin" and "sys"
It's less weird to me to just put everything in one header and have that be "matrix things" instead. Or possibly split it up in chunks if there's a sensible way to do so (like split the chunk working with a mtx stack)

Now if the goal isn't usability, what is it?

Comment (cadmic) [reply include/matrix.h:45]: I mean it's evidence but not the main reason, if the two files worked better with each other (or other factors like both being named sys_ or z_) then I would consider them more the same system. The main reason is that the code does not treat sys_matrix and skin_matrix as a unified whole.

For example, `Matrix_TranslateRotateZYX` and `SkinMatrix_SetTranslateRotateZYX` theoretically do the same thing, but actually are pretty different. Not just matrix stack vs no stack, but also different parameter types and different implementations that share no code and probably have different rounding behavior (though I haven't checked). So I think it would be confusing to mix them too closely

Comment (cadmic) [reply include/matrix.h:45]: > Now if the goal isn't usability, what is it?

Documentation and reverse-engineering I guess? Like I want to readers the impression of "oh huh, there's two different matrix subsystems and they're often completely different, that's cool/horrifying/good-to-know".

I don't think it's less usable, if you're modding then you should probably just use sys_matrix and forget about skin_matrix entirely

Comment (fig02) [reply include/matrix.h:45]: Think im personally bowing out of this conversation for now, cause I'd just be repeating myself. I've said what I want to say.
Would love to hear more thoughts than just us 3, if anyone else wants to give their opinion.

And lastly I'll say that if we don't come to a conclusion on this now, that we should go back to 1 header until we do decide

Comment (Dragorn421) [reply include/matrix.h:45]: > Documentation and reverse-engineering I guess? Like I want to give readers the impression of \"oh huh, there's two different matrix subsystems and they're often completely different, that's cool/horrifying/good-to-know\

Alright that's reasonable to me (though I dislike splitting the systems based on the .c. During development I could see a dev adding a matrix function they need to either file, whichever they come across first)I'll switch to sys_matrix.h + skin_matrix.h

Comment (AngheloAlf) [reply include/matrix.h:45]: Personally I prefer if the headers were separated by their systems.
In this specific case the systems would be named after their C files, yes. That's because it happens to be organized this way. But this doesn't apply to most systems, like all the different systems present on z_actor.c.

So I agree with you dragorn that we should split headers by system, this just happens to more or less match the corresponding C files. Since there's currently no consensus on how to name the headers I think we should name them after the C files since that's the most obvious way, we can always rename them later.

Since you are concerned about usability then you could add a comment at the top of the header explaining the system a bit and how it differentiates to the other matrix header.

tl;dr: I prefer sys_matrix.h + skin_matrix.h and docs comment on each of them.

**Review by fig02 — APPROVED** [review]

**Review by AngheloAlf — APPROVED** [review]

File: include/skin_matrix.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#ifndef MATRIX_H
+#define MATRIX_H
```
Comment (cadmic) [inline include/skin_matrix.h:2]: include guard name doesn't match

File: include/sys_matrix.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
+#ifndef MATRIX_STACK_H
+#define MATRIX_STACK_H
```
Comment (cadmic) [inline include/sys_matrix.h:2]: ditto

**Review by cadmic — APPROVED** [review]

**Comment by mzxrules** [comment]
PR #1972 also conflicts with this because I implemented my own [sys_matrix.h](https://github.com/zeldaret/oot/blob/e6c7ef01558e476e767ccefe603ec8ee812df2cf/include/sys_matrix.h) that mimics MM's own [sys_matrix.h](https://github.com/zeldaret/mm/blob/main/include/sys_matrix.h)

I'm fine with dropping the inclusion of sys_matrix.h in #1972 in favor of this pr, though I would prefer the two headers stay mostly consistent between games.

**Comment by Dragorn421** [comment]
I don't know where to take this
